Convert manifests + BSL api client to kubebuilder

carlisia opened the pull request:
- All CRD manifests, and generation of manifests generated by kubebuilder (same but different)
- BSL api client code generation ported from go-client to kubebuilder
- Substituted all BSL usage to the kubebuilder api client
- Tied in CRD/code generation with existing scripts

Closes: https://github.com/vmware-tanzu/velero/issues/2604

TODO:
- [x] basic CRUD tests for backup and backup storage location
- [x] more manual tests, esp. with restic

https://www.loom.com/share/b96ce9e9cd3846428ef4243ed9a1828a

carlisia commented:
Rebasing this PR is not fun @vmware-tanzu/velero-maintainers!

carlisia commented on api/v1/backupstoragelocation_types.go:
```diff
@@ -21,33 +21,85 @@ import (
+// +kubebuilder:object:generate=true
+// +kubebuilder:storageversion
+// +kubebuilder:subresource:status
+// +kubebuilder:printcolumn:name="Phase",type="string",JSONPath=".status.phase",description="Backup Storage Location status such as Available/Unavailable"
```
This kubectl column would be empty now but filled in once this goes in: https://github.com/vmware-tanzu/velero/pull/2493

skriss commented on api/v1/backupstoragelocation_types.go:
```diff
@@ -1,5 +1,5 @@
 /*
-Copyright 2018 the Velero contributors.
+Copyright 2018, 2020 the Velero contributors.
```
Thinking through the implications of our API types living in two separate packages, and also if we need to consider this a breaking change. It will be breaking for anyone who vendors our code to use the API/client. That said, we haven't historically actually maintained backwards compatibility in our exported API surface..

skriss commented on hack/update-generated-crd-code.sh:
```diff
@@ -21,32 +21,4 @@ set -o xtrace
-  paths=./pkg/apis/velero/v1/...
-
-go generate ./pkg/generated/crds
+go generate ./config/crd/crds
```
I think we need to keep most of the code that was removed here, for the non-kubebuilder API types/clients/etc

carlisia replied:
Hm. 🤔 I think this will work. Here's how I tested it: added a field to `pkg/apis/velero/v1/backup.go` and ran `make update`, and the other 3 files below got updated as expected. If there's nothing missing, then I think it's good like this?

![image](https://user-images.githubusercontent.com/16508/83192457-68d80380-a0ea-11ea-9e71-499407da5205.png)

carlisia replied:
Although all api types (besides the BSL) remain where they were, their corresponding manifest files and the raw CRDs are being generated in the same way and placed in the same places.

skriss replied:
There's also the generated clients/listers/informers in `pkg/generated`, and the deep-copy functions that go directly in `pkg/apis/velero/v1/zz_generated.deepcopy.go`, that potentially need to update when one of the existing API types updates. I think specifically [this code](https://github.com/vmware-tanzu/velero/pull/2561/files/bf5052d695c4617f97bd62df2d52bfb88d4163ab#diff-892d90adfc5482f1fee4517a4ad27aeeL38-L45) needs to be retained to properly update those.

skriss replied:
And I guess [this code](https://github.com/vmware-tanzu/velero/pull/2561/files/bf5052d695c4617f97bd62df2d52bfb88d4163ab#diff-892d90adfc5482f1fee4517a4ad27aeeL24-L31) as well.

carlisia replied:
Hmyeah... I totally lost track of this. Will fix.

ashish-amarnath replied:
to confirm this works fully, I would suggest deleting all the generated code and then run the code generation step again to confirm everything comes backup.

skriss commented on main.go:
```diff
@@ -0,0 +1,18 @@
+/*
```
can this be removed?

carlisia replied:
The file? Yes and no :)

Yes, we can remove it and everything will work fine.

Things will not work fine IF/WHEN we try to create a new scaffold for either a resource or a controller using `kubebuilder`. `kubebuilder` will halt the generation of files when it doesn't find `main.go`. Well, we don't do this everyday (generate new resources/controllers) so (again, if we go in this direction) we COULD just remember to add a `main.go` file while doing that and remove it again.

skriss commented on pkg/backup/request.go:
```diff
@@ -20,6 +20,7 @@ import (
 	"fmt"
 	"sort"
 
+	veleroapiv1 "github.com/vmware-tanzu/velero/api/v1"
```
if we do go ahead with two separate packages for the API types, we need to come up with a clearer way to alias them :)

carlisia replied:
Yes, I'm very open to have a better alias name. For lack of a great idea, I turned `velero/api/v1` into `veleroapiv1`, which (obvs) follows the order of the directories (which doesn't mean is great but has a meaning), and also was NOT the `velerov1api` which we use for our existing api package.

skriss commented on pkg/controller/suite_test.go:
```diff
@@ -0,0 +1,36 @@
+	// +kubebuilder:scaffold:imports
+)
+
+func newFakeClient(g *WithT, initObjs ...runtime.Object) client.Client {
```
IDK how much of an issue this would cause, but I'd prefer not to introduce the gomega dependency if we don't need to - we're already using stretchr/testify for the `assert` and `require` libs everywhere else.

carlisia replied:
Not needed, you're right. I'll remove it! If we go in this direction, we can consider if we want to continue using the same testing packages, and have the controllers look like they do today, or use the BDD style and have them look like this: https://github.com/kubernetes-sigs/cluster-api/blob/master/controllers/cluster_controller_phases_test.go.

carlisia commented on pkg/cmd/cli/restic/server.go:
```diff
@@ -130,6 +152,21 @@ func newResticServer(logger logrus.FieldLogger, factory client.Factory) (*restic
+		return nil, err
+	}
+
+	ctrl.SetLogger(zap.New(zap.UseDevMode(true)))
```
Oops, I should have removed that and used ours. Fixing.

carlisia commented on pkg/cmd/cli/backup/create.go:
```diff
@@ -33,10 +38,23 @@ import (
 
 const DefaultBackupTTL time.Duration = 30 * 24 * time.Hour
 
+var (
+	scheme = runtime.NewScheme()
+)
+
+func init() {
+	_ = clientgoscheme.AddToScheme(scheme)
+
+	_ = velerov1api.AddToScheme(scheme)
+	// +kubebuilder:scaffold:scheme
+}
```
Maybe @ashish-amarnath can assist me here. When I only had it in the api package, it didn't work. The cli code didn't recognize any scheme.

skriss replied:
OK, I can try to dig into it some more as well.

ashish-amarnath replied:
we can look at examples to figure out how to consolidate this in one place.

ashish-amarnath commented on Makefile:
```diff
@@ -40,6 +40,14 @@ CLI_PLATFORMS ?= linux-amd64 linux-arm linux-arm64 darwin-amd64 windows-amd64 li
 MANIFEST_PLATFORMS ?= amd64 ppc64le arm arm64
 
+#kubebuilder automatically runs "make" when a resource or controller is created
+all: generate-all
```
This is just creating an alias for `generate-all`.
You can undo this change and keep `generate-all` as the first target, if that is you wanted.
You can also undo the change on line #92

ashish-amarnath commented on api/v1/backupstoragelocation_types.go:
```diff
@@ -21,33 +21,85 @@ import (
+	AccessMode BackupStorageLocationAccessMode `json:"accessMode,omitempty"`
+}
+
+// +kubebuilder:object:root=true
```
should we also add the annotation to generate RBAC?

carlisia replied:
@ashish-amarnath I really don't know. The original BSL didn't have that marker. Please let me know if you think I should add it, but also if you could help me understand why it'd be great.

ashish-amarnath replied:
KubeBuilder is able to generate RBAC for the controllers to run wit, based on the annotation. The question was whether we want to use that. We can table that discussion for when we move all the controllers.

ashish-amarnath commented on api/v1/backupstoragelocation_types.go:
```diff
@@ -117,31 +148,3 @@ const (
 
 // TODO(2.0): remove the AccessMode field from BackupStorageLocationStatus.
```
should lines 149 and 150 also be removed?

carlisia replied:
No, they have nothing to do with this PR, those are slated to be done for v2.0.

ashish-amarnath commented on config/crd/bases/velero.io_backups.yaml:
```diff
@@ -14,7 +14,6 @@ spec:
     listKind: BackupList
     plural: backups
     singular: backup
-  preserveUnknownFields: false
```
in all CRDs, throughout.

nrb replied:
I _think_ it would be `false` if not specified, since the empty value for bools in Go is `false`.

ashish-amarnath commented on config/certmanager/certificate.yaml:
```diff
@@ -0,0 +1,26 @@
+# The following manifests contain a self-signed issuer CR and a certificate CR.
```
this is most likely coming in as part of the validation webhook that comes with kubebuilder

ashish-amarnath commented on config/samples/velero_v1_backupstoragelocation.yaml:
```diff
@@ -0,0 +1,7 @@
+  name: backupstoragelocation-sample
+spec:
+  # Add fields here
+  foo: bar
```
this example needs to be completed

ashish-amarnath commented on pkg/cmd/cli/restic/server.go:
```diff
@@ -157,6 +195,12 @@ func (s *resticServer) run() {
 	var wg sync.WaitGroup
 
+	// TODO(carlisia): how to handle the fetching of the bsl informer:
+	// - options are: 1) get informer for specific obj (below, but w/o namespace or obj name info because we don't know it here) or 2) for a specific kind, which would be CRD?
```
option 2 feels more correct.
Were you not able to do that?

nrb commented on config/default/kustomization.yaml:
```diff
@@ -0,0 +1,70 @@
+# Adds namespace to all resources.
+namespace: velero-system
```
I think this namespace would need to change, wouldn't it?

nrb commented on config/prometheus/kustomization.yaml:
```diff
@@ -0,0 +1,2 @@
+resources:
```
Do we want prometheus coming in with this change? This question applies to all the files in the `config/prometheus` directory.

nrb commented on pkg/cmd/server/server.go:
```diff
@@ -100,6 +110,10 @@ const (
 )
 
+var (
+	scheme = runtime.NewScheme()
```
Can you add some comments to explain why this and the init function are doing scheme calls? It's not clear to me why this is happening in the server.go file.

carlisia commented:
Deleting the entire `generated` directory and running `make update`
works. Modifying an api and running `make verify` works as expected.

Pending: addressing questions and maybe a couple clean ups as per the reviews so far, but no big change expected.

carlisia commented:
@vmware-tanzu/velero-maintainers PTAL when you have a chance, it's ready for review.

There are 3 //todo(carlisia) to call attention to some questions I have about how to best refactor those code sections.

All other reviews were addressed, minus one question for Ashish (see above).

skriss commented:
sorry i've been slow here - planning to look at this tomorrow morning!

carlisia commented:
WIP BSL controller: https://github.com/vmware-tanzu/velero/pull/2617

carlisia commented on Makefile:
```diff
@@ -296,3 +296,35 @@ serve-docs:
+GOBIN=$(shell go env GOBIN)
+endif
+
+# Generate manifests e.g. CRD, RBAC etc.
+manifests: controller-gen
+	$(CONTROLLER_GEN) $(CRD_OPTIONS) rbac:roleName=manager-role webhook paths="./pkg/apis/velero/v1" output:crd:artifacts:config=config/crd/bases
+
+# find or download controller-gen
+# download controller-gen if necessary
+controller-gen:
+ifeq (, $(shell which controller-gen))
+	@{ \
+	set -e ;\
+	CONTROLLER_GEN_TMP_DIR=$$(mktemp -d) ;\
+	cd $$CONTROLLER_GEN_TMP_DIR ;\
+	go mod init tmp ;\
+	go get sigs.k8s.io/controller-tools/cmd/controller-gen@v0.2.5 ;\
+	rm -rf $$CONTROLLER_GEN_TMP_DIR ;\
+	}
+CONTROLLER_GEN=$(GOBIN)/controller-gen
+else
+CONTROLLER_GEN=$(shell which controller-gen)
+endif
```
Ohh. No, no grand plan not to use the image, and I think using the image is probably best. I'll fix this.

carlisia commented on pkg/apis/velero/v1/register.go:
```diff
@@ -22,20 +22,6 @@ import (
-)
-
-// GroupName is the group name for the Velero API
-const GroupName = "velero.io"
```
It was being used, but I changed to use it directly and also that code is now in `groupgversion_info.go`:
```
SchemeGroupVersion = schema.GroupVersion{Group: "velero.io", Version: "v1"}
```

carlisia commented:
> BTW I'm happy to spend some time testing this if that'd be helpful; let me know if there are any specific areas you'd like me to focus on!

Oh if you could run some manual tests with Restic it would be awesome!

carlisia commented on pkg/restic/repository_manager.go:
```diff
@@ -79,20 +82,19 @@ type RestorerFactory interface {
+	secretsLister      corev1listers.SecretLister
+	repoLister         velerov1listers.ResticRepositoryLister
+	repoInformerSynced cache.InformerSynced
+	kbCache            kbcache.Cache
```
That is correct. I think I had added this before I understood that get/list come from cache when using the client, then forgot to change it. Changing it now.

carlisia replied:
Oh, actually, no. I used it so I would have access to `WaitForCacheSync`, which the client doesn't provide. I think we need this check? Please let me know.

https://github.com/vmware-tanzu/velero/blob/84698c88ff22229434fa3da6d8cbd99b9a4c7def/pkg/restic/repository_manager.go#L273

skriss replied:
so this is very related to https://github.com/vmware-tanzu/velero/pull/2561#discussion_r438242022, and I'm not 100% confident I understand the controller-runtime client behavior here. My sense was that it combines a ~lister/cache and an actual API client behind the scenes, so it'll return an object from the cache if it can, else it'll hit the API.

If that's the case, then we *may* not need to wait for the cache sync here, since we can use the client regardless. But again, i'm not totally sure I've got the mechanics of the new client down. Perhaps @ashish-amarnath can weigh in here.

ncdc replied:
If you are using something that looks like this to create a client:

https://github.com/kubernetes-sigs/cluster-api/blob/714b0a97484465a05a67d8ccd1b64aa388b9fac2/util/util.go#L636-L646

Then any calls to `client.Get()` will not return any data until after the cache is synced.

Additionally, it will only return data from the cache. It does not fall back to doing a live read against the apiserver.

Does this help? Happy to hop on a call if you need it.

shomron replied:
I don't have all the context around this change, but I'll try to weigh in :) Internally `controller-runtime`'s cache will `WaitForCacheSync` when it tries to access any of the informers it is managing, so I'm not sure why you'd have to call it directly.

carlisia replied:
I think that clarifies the question, @ncdc, thank you!

shomron replied:
> My sense was that it combines a ~lister/cache and an actual API client behind the scenes, so it'll return an object from the cache if it can, else it'll hit the API.

@skriss that's not quite my understanding.
If you use the default _Client_ setup by the controller-runtime Manger, it will delegate to the informer cache for typed resource reads and pass-through to the API Server for unstructured resource reads. For writes it will always pass-through to the API Server.

That is wired up [here](https://github.com/kubernetes-sigs/controller-runtime/blob/32b4434331e72fed344f71d278982827e08db2a8/pkg/manager/manager.go#L347).

If you use the _Cache_, you should be hitting all informers all the time.

ncdc replied:
Note, in Cluster API, we explicitly made the choice to use a cache for all resources (typed and unstructured).

skriss replied:
thanks @shomron and @ncdc, that's helpful clarification

carlisia replied:
Yes, what @shomron said is what I learned too. In other words, if the cache isn't started for example, the get/list will fail. Ways around that are, specifically pass an unstructured obj, or call the `GetAPIReader()`.

ncdc replied:
If you use the controller-runtime `Manager`, it auto-starts all the caches for all the kinds you care about (have reconcilers for) automatically, so you generally don't need to worry about caches not being started in that setup.

carlisia replied:
Hm. That's a good point. At this point there is no reconciler and no manager, we only have the resource.

carlisia replied:
No actually, that's wrong, we do have a manager, just not any reconciler.

skriss commented:
sure, I'll try to spend some time on that this afternoon!

nrb commented on pkg/cmd/cli/restic/server.go:
```diff
@@ -130,6 +143,21 @@ func newResticServer(logger logrus.FieldLogger, factory client.Factory) (*restic
+		return nil, err
+	}
+
+	// +kubebuilder:scaffold:builder
```
Honestly having a hard time finding it in https://github.com/kubernetes-sigs/kubebuilder/search?q=%2Bkubebuilder%3Ascaffold%3Abuilder&type=Code, too. I'm only seeing it in example code without much explanation of what it's actually doing.

ncdc replied:
It's a placeholder so kubebuilder knows where to do code insertions.

nrb replied:
@ncdc Thanks. Does it map to [this method](https://github.com/kubernetes-sigs/kubebuilder/blob/6e0d3d4f18ecdc335177d28e1d6cd4d7b2930242/pkg/plugin/internal/machinery/scaffold.go#L66)?

carlisia replied:
Oh. So I guess we can deleted bc kubebuilder always writes to `main.go` and would never write to this file.

ncdc replied:
@nrb I don't know (I'm not familiar with how it's implemented in kubebuilder).

How did this line come to exist in server.go?

ashish-amarnath replied:
I don't know what code this marker generates. But I think we don't need any kubebuilder code generated here and that this may be removed.

carlisia commented:
@vmware-tanzu/velero-maintainers addressed all code reviews, PTAL!

skriss commented on pkg/controller/pod_volume_backup_controller.go:
```diff
@@ -42,19 +42,23 @@ import (
+	podLister              corev1listers.PodLister
+	pvcLister              corev1listers.PersistentVolumeClaimLister
+	pvLister               corev1listers.PersistentVolumeLister
+	backupLocationInformer k8scache.Informer
```
I think we can drop the `backupLoationInformer` and associated import now

carlisia commented on pkg/cmd/cli/restic/server.go:
```diff
@@ -43,6 +46,13 @@ import (
+)
+
+var (
+	scheme = runtime.NewScheme()
```
Good catch. Adding it right before creating the manager.

skriss commented:
I'm seeing a couple of errors in the velero server log (haven't gotten to restic yet) on startup:

```
time="2020-06-11T21:20:28Z" level=error msg="Error getting backup storage locations from lister" controller=backup-sync error="the cache is not started, can not read objects" error.file="/github.com/vmware-tanzu/velero/pkg/controller/backup_sync_controller.go:129" error.function="github.com/vmware-tanzu/velero/pkg/controller.(*backupSyncController).run" logSource="pkg/controller/backup_sync_controller.go:129"
...
time="2020-06-11T21:21:28Z" level=error msg="Error patching backup location's last-synced time" backupLocation=default controller=backup-sync error="backupstoragelocations.velero.io \"default\" not found" error.file="/github.com/vmware-tanzu/velero/pkg/controller/backup_sync_controller.go:311" error.function="github.com/vmware-tanzu/velero/pkg/controller.(*backupSyncController).run" logSource="pkg/controller/backup_sync_controller.go:311"
```

The BSL definitely exists.

ncdc commented:
If/when you convert to use the controller-runtime Manager for everything, it auto-starts all the caches. Otherwise, you'll need to start them manually before trying to use the client.

carlisia commented:
Ohh. Right. I'll add a start for the manager.

carlisia commented:
I'm getting this and the other PR that adds the BSL controller confused.

I had already added the start. @ncdc do you mean something different from this: https://github.com/vmware-tanzu/velero/blob/ed6fb4568db0bbd771196f28344dc2cb1deda8a2/pkg/cmd/server/server.go#L911

ncdc commented:
@carlisia `s.mgr.Start()` is what I meant. I'm guessing what you're seeing is a race between your controllers (pre-existing / outside of the Manager's control) and the Manager starting the caches. If you complete the massive effort to convert 100% of Velero to be controller-runtime based, this issue will go away. But while you have a mix of controller-runtime Reconcilers and "legacy" Velero controllers, you'll have to wait for caches to sync before trying to use the client. Does that make sense?

ashish-amarnath commented on hack/update-all.sh:
```diff
@@ -24,3 +24,5 @@ for f in ${HACK_DIR}/update-*.sh; do
   fi
   $f
 done
+
```
changes to this file may be reverted

carlisia commented:
@ncdc yes, very clear. I'm glad you jumped in this PR. I did suspect that we would have to wait, but didn't know this will go away once all controllers are converted (eventually). I guess we'll have to add a wait there...

ashish-amarnath commented on pkg/cmd/server/server.go:
```diff
@@ -439,8 +463,10 @@ func (s *server) validateBackupStorageLocations() error {
-	locations, err := s.veleroClient.VeleroV1().BackupStorageLocations(s.namespace).List(metav1.ListOptions{})
-	if err != nil {
+	locations := &velerov1api.BackupStorageLocationList{}
+	if err := s.mgr.GetAPIReader().List(context.Background(), locations, &kbclient.ListOptions{
```
Is there a reason why we are not using `GetClient`? The `GetAPIReader` talks directly to the API server and does not use the cache.

carlisia replied:
Yes, at this point the cache hasn't been started yet.

ashish-amarnath replied:
suggest adding a comment to that effect.

nrb replied:
👍 to a comment for explanation on `GetAPIReader`. We went over this on a call, but I'm likely to forget in a week 😄

carlisia replied:
Done.

ashish-amarnath commented on pkg/cmd/server/server.go:
```diff
@@ -880,6 +904,15 @@ func (s *server) runControllers(defaultVolumeSnapshotLocations map[string]string
+	wg.Add(1)
+	go func() {
+		defer wg.Done()
+		// +kubebuilder:scaffold:builder
```
this kubebuilder marker can also be removed.

ashish-amarnath commented on pkg/controller/backup_sync_controller.go:
```diff
@@ -101,35 +100,38 @@ func NewBackupSyncController(
 // followed by the rest of the locations in no particular order.
-func orderedBackupLocations(locations []*velerov1api.BackupStorageLocation, defaultLocationName string) []*velerov1api.BackupStorageLocation {
-	var result []*velerov1api.BackupStorageLocation
+func orderedBackupLocations(locationList *velerov1api.BackupStorageLocationList, defaultLocationName string) []velerov1api.BackupStorageLocation {
```
Changes to this function may not be necessary.
We can call this function like
```
locations := orderedBackupLocations(locationList.Items, c.defaultBackupLocation)
```

carlisia replied:
`locationList.Items` is a `[]velerov1api.BackupStorageLocation`, and `orderedBackupLocations` needs a `[]&velerov1api.BackupStorageLocation` because of things like this: `result = append(result, locations[:i]...)`.

ashish-amarnath commented on pkg/restic/common.go:
```diff
@@ -229,17 +233,20 @@ func TempCACertFile(caCert []byte, bsl string, fs filesystem.Interface) (string,
 	}
 
-	return nil, nil
+	return location, nil
```
Considering that this function is called `GetCACert`, it should return the CACert associated with the BSL and not the BSL object itself.

ashish-amarnath replied:
If we need a function to get the BSL, suggest creating a new one to do that.

carlisia replied:
Yes, very good point!

ashish-amarnath commented on pkg/restic/common.go:
```diff
@@ -254,12 +261,7 @@ func NewPodVolumeRestoreListOptions(name string) metav1.ListOptions {
-		return nil, errors.Wrap(err, "error getting backup storage location")
-	}
-
+func AzureCmdEnv(loc *velerov1api.BackupStorageLocation) ([]string, error) {
```
Changes in this file are all API breaking changes.
For purposes of backwards compatibility, we should create new functions and add a deprecation notice to the old ones.

ashish-amarnath commented on site/docs/master/code-standards.md:
```diff
@@ -43,7 +43,8 @@ Example:
     	corev1listers "k8s.io/client-go/listers/core/v1"
-        
+       
+        veleroapiv1 "github.com/vmware-tanzu/velero/api/v1" // for the new kubebuilder apis 
         velerov1api ""github.com/vmware-tanzu/velero/pkg/apis/velero/v1"
```
`velerov1api` is a reused import alias.
Also looks like line#48 has a compiler error: unmatched `"`.

carlisia replied:
`github.com/vmware-tanzu/velero/api/v1` no longer exists so this change is being reverted.

carlisia commented:
Addressed all code reviews.

No more caching problems: Fixed it by registering all existing (non runtime-controller) controllers to the manager: https://github.com/vmware-tanzu/velero/blob/cd048ee78d5c0825c08dc51314999051a5b37b69/pkg/cmd/server/server.go#L897

ashish-amarnath commented on pkg/cmd/cli/backup/create.go:
```diff
@@ -142,13 +145,22 @@ func (o *CreateOptions) Validate(c *cobra.Command, args []string, f client.Facto
 		return err
 	}
 
+	client, err := f.KubebuilderClient()
```
should the
```
client, err := f.Client()
```
in other functions also be updated to use the `KubebuilderClient()`?

do we still want to continue having the field
```
client veleroclient.Interface
```
in `CreateOptions` Line#99)

carlisia replied:
The answers are `no` and `yes`. This is because this PR only changes the api client for the BSL resource. So if you want to handle, say, backup resources, the client is not changing.

ashish-amarnath commented on pkg/client/factory.go:
```diff
@@ -137,6 +143,21 @@ func (f *factory) DynamicClient() (dynamic.Interface, error) {
 	return dynamicClient, nil
 }
 
+func (f *factory) KubebuilderClient() (kbclient.Client, error) {
```
`VeleroV1APIClient` seems like a better name for this function. WDYT?

carlisia replied:
I'm not sure. Looking at the other clients, none has version in the name:
![image](https://user-images.githubusercontent.com/16508/85134006-a2cf8d80-b1f0-11ea-9ed7-c20079cb595c.png)

`Client` would be the best, but it's already in use so we need something to differentiate this new kubebuilder client.

@nrb do you have any opinion on this?

nrb replied:
Looking through what's returned, `VeleroV1APIClient` doesn't quite make sense to me, since we get https://pkg.go.dev/sigs.k8s.io/controller-runtime/pkg/client?tab=doc#Client, which is a generic client that can interact with anything in the Kubernetes API server, so it's not really just for Velero V1; we just happen to be using it for that now. Really, we could register any API group/scheme to it and use this client to get them.

In the current file, `Client` returns https://pkg.go.dev/github.com/vmware-tanzu/velero@v1.4.0/pkg/generated/clientset/versioned?tab=doc#Interface, which has the `VeleroV1()` method, which makes it a typed and specific client.

I'm leaning towards keeping the `KubeBuilderClient` name for now, then when we complete the transition, renaming it to `KubeClient`, which would replace the existing `KubeClient` for API server interactions. That's a much further-looking plan, though.

Thoughts?

ashish-amarnath commented on pkg/cmd/cli/restic/server.go:
```diff
@@ -182,26 +203,27 @@ func (s *resticServer) run() {
+		s.logger.Fatal("Problem starting manager", err)
+	}
 
 	s.logger.Info("Waiting for all controllers to shut down gracefully")
```
`s.mgr.Start` is a blocking call, so this message wouldn't be logged until `Start` returns.
I would suggest removing this log message.

carlisia replied:
Good point!

ashish-amarnath commented on pkg/cmd/server/server.go:
```diff
@@ -867,23 +892,20 @@ func (s *server) runControllers(defaultVolumeSnapshotLocations map[string]string
+		s.logger.Fatal("Problem starting manager", err)
+	}
 
 	s.logger.Info("Waiting for all controllers to shut down gracefully")
```
same comment as https://github.com/vmware-tanzu/velero/pull/2561/files?file-filters%5B%5D=.go&file-filters%5B%5D=.md&file-filters%5B%5D=.mod&file-filters%5B%5D=.yaml&file-filters%5B%5D=No+extension#r442545437

ashish-amarnath commented on pkg/cmd/server/server.go:
```diff
@@ -867,23 +892,20 @@ func (s *server) runControllers(defaultVolumeSnapshotLocations map[string]string
+	s.logger.Info("Server starting...")
 
-	<-ctx.Done()
+	if err := s.mgr.Start(s.ctx.Done()); err != nil {
```
`err` should be returned

carlisia replied:
I'm doing a `s.logger.Fatal` so I don't think it's needed?

nrb replied:
Yeah, I would think if the error's handled, it shouldn't also be returned.

ashish-amarnath commented on pkg/cmd/util/output/backup_storage_location_printer.go:
```diff
@@ -30,7 +30,7 @@ var (
 		{Name: "Provider"},
 		{Name: "Bucket/Prefix"},
-		{Name: "Status"},
+		{Name: "Phase"},
```
I think `Status` is a better name.

carlisia replied:
To clarify: I changed this (seemly unrelated to this PR) field to match the new column `Phase` we will be displaying with `kubectl`:

![image](https://user-images.githubusercontent.com/16508/85134582-97309680-b1f1-11ea-9e86-cf20fb8b00ea.png)

Pls ignore the 2 in `PHASE2`, I haven't refreshed my cluster.

The reason why I chose `phase` instead of `status` is because the field is `status.phase`, so this is more accurate and also leaves room to add other `status.x` fields if we want to.

nrb replied:
I like `Phase` here, since `Status` is the name of a field that contains other fields.

It also directly maps to the field's name.

ashish-amarnath commented on pkg/controller/download_request_controller_test.go:
```diff
@@ -27,7 +27,10 @@ import (
 	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
 	"k8s.io/apimachinery/pkg/util/clock"
 
+	"sigs.k8s.io/controller-runtime/pkg/client"
```
this import may be combined with the above import block.

ashish-amarnath commented on pkg/controller/pod_volume_backup_controller.go:
```diff
@@ -228,10 +229,11 @@ func (c *podVolumeBackupController) processBackup(req *velerov1api.PodVolumeBack
 
 	// if there's a caCert on the ObjectStorage, write it to disk so that it can be passed to restic
-	caCert, err := restic.GetCACert(c.backupLocationLister, req.Namespace, req.Spec.BackupStorageLocation)
+	caCert, err := restic.GetCACert(c.kbClient, req.Namespace, req.Spec.BackupStorageLocation)
```
Would it be better to get the BSL object once and pass it to the subsequent functions?
This would require creating new APIs for this.
WDYT?

carlisia replied:
That was how the last version of this PR was. I'd rather not make any more changes to this. This is as close as the original as it can be.

carlisia replied:
Not the last version actually, but a couple versions ago.

ashish-amarnath replied:
Yeah I agree the API was changed to accept a BSL object and that was a breaking API change.
I was hoping for a new API, not replacing the one that exists, to be created that works on the BSL object directly.

I think this change is still breaking the API. In that, I mean, if someone using this API in their code were to upgrade their dependency to a version of velero with this change, then their code wouldn't compile. No?

carlisia replied:
Yes, it'll be a breaking change. I talked to @nrb about this, this is not an area of the code anyone should be importing. We'are also going to start moving things to inside a (to be added) `internal` package. Are you ok with this breaking change?

ashish-amarnath replied:
As long as none of our plugins are using these functions, we should be OK.
Ideally we should be deprecating APIs across releases.

nrb replied:
@ashish-amarnath Yeah, that would be ideal, though we've got a lot of examples of where we move our code around without deprecating.

I think I'm ok with this change, personally, even though it is breaking with strict semantic versioning. We have a lot of "accidentally" public code that we've been pretty rude about moving around in the past, and I think it would be good to document our true, intentional public API.

Carlisia made https://github.com/vmware-tanzu/velero/issues/2654 to do this, and we'll ask folks to comment there as we move things into `internal`. If moving something breaks someone, we can revisit.

This definitely isn't ideal, but we've also got to be able to move forward.

ashish-amarnath commented on pkg/controller/restore_controller.go:
```diff
@@ -48,6 +49,8 @@ import (
 	kubeutil "github.com/vmware-tanzu/velero/pkg/util/kube"
 	"github.com/vmware-tanzu/velero/pkg/util/logging"
+
+	"sigs.k8s.io/controller-runtime/pkg/client"
```
this import should be part of the import block from L31-37

ashish-amarnath commented on pkg/controller/backup_deletion_controller_test.go:
```diff
@@ -34,8 +35,10 @@ import (
+	"sigs.k8s.io/controller-runtime/pkg/client"
 
 	velerov1 "github.com/vmware-tanzu/velero/pkg/apis/velero/v1"
+	velerov1api "github.com/vmware-tanzu/velero/pkg/apis/velero/v1"
```
this is already imported as `velerov1` on line 40

ashish-amarnath commented on pkg/controller/download_request_controller_test.go:
```diff
@@ -27,7 +27,10 @@ import (
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
 	v1 "github.com/vmware-tanzu/velero/pkg/apis/velero/v1"
+	velerov1api "github.com/vmware-tanzu/velero/pkg/apis/velero/v1"
```
looks like something is still using the `v1.`
Once we've removed all the v1 import alias, we move the double-import.

ashish-amarnath replied:
This still needs to be addressed.

ashish-amarnath commented on pkg/controller/gc_controller_test.go:
```diff
@@ -32,6 +34,8 @@ import (
 	"k8s.io/apimachinery/pkg/watch"
 	core "k8s.io/client-go/testing"
 
+	kbclient "sigs.k8s.io/controller-runtime/pkg/client"
```
can be combined with the import block from L-28-35.

ashish-amarnath commented on pkg/controller/gc_controller_test.go:
```diff
@@ -23,6 +23,8 @@ import (
 	"testing"
 	"time"
 
+	velerov1api "github.com/vmware-tanzu/velero/pkg/apis/velero/v1"
```
double-import from L-39
and can be moved to L39

carlisia replied:
Fixed.

ashish-amarnath commented on pkg/controller/pod_volume_backup_controller.go:
```diff
@@ -42,6 +42,8 @@ import (
 	"github.com/vmware-tanzu/velero/pkg/util/filesystem"
 	"github.com/vmware-tanzu/velero/pkg/util/kube"
+
+	"sigs.k8s.io/controller-runtime/pkg/client"
```
can be combined with import block from L-26-36

ashish-amarnath commented on pkg/controller/restic_repository_controller.go:
```diff
@@ -31,19 +32,21 @@ import (
 	velerov1listers "github.com/vmware-tanzu/velero/pkg/generated/listers/velero/v1"
 	"github.com/vmware-tanzu/velero/pkg/restic"
+
+	"sigs.k8s.io/controller-runtime/pkg/client"
```
same comment as the one [here](https://github.com/vmware-tanzu/velero/pull/2561/files?file-filters%5B%5D=.go&file-filters%5B%5D=.md&file-filters%5B%5D=.mod&file-filters%5B%5D=.yaml&file-filters%5B%5D=No+extension#r442559843)

ashish-amarnath commented on pkg/controller/restore_controller_test.go:
```diff
@@ -18,11 +18,14 @@ package controller
 	"testing"
 	"time"
 
+	velerov1api "github.com/vmware-tanzu/velero/pkg/apis/velero/v1"
```
double-import from L-40.
Can be moved to L-40

carlisia replied:
Fixed.

ashish-amarnath commented on pkg/install/resources.go:
```diff
@@ -19,16 +19,17 @@ package install
 import (
 	"time"
 
+	velerov1api "github.com/vmware-tanzu/velero/pkg/apis/velero/v1"
```
can be moved to L-33

ashish-amarnath commented on pkg/restic/common.go:
```diff
@@ -17,11 +17,15 @@ limitations under the License.
 	"strings"
 	"time"
 
+	"sigs.k8s.io/controller-runtime/pkg/client"
```
group imports with import block from L-29-32

carlisia replied:
I didn't think we were manually fussing with where imports go?

ashish-amarnath commented on pkg/restic/common_test.go:
```diff
@@ -17,18 +17,22 @@ limitations under the License.
 	"testing"
 
-	velerov1listers "github.com/vmware-tanzu/velero/pkg/generated/listers/velero/v1"
+	"k8s.io/apimachinery/pkg/runtime"
```
combine with import block from L-27-35

ashish-amarnath commented on site/docs/master/code-standards.md:
```diff
@@ -43,8 +43,8 @@ Example:
-        
-        velerov1api ""github.com/vmware-tanzu/velero/pkg/apis/velero/v1"
+       
+        velerov1api "github.com/vmware-tanzu/velero/pkg/apis/velero/v1"
```
This change should be copied over to the docs for the other versions as well.

ashish-amarnath commented on site/docs/v1.3.1/code-standards.md:
```diff
@@ -43,7 +43,7 @@ Example:
     	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
     	corev1client "k8s.io/client-go/kubernetes/typed/core/v1"
     	corev1listers "k8s.io/client-go/listers/core/v1"
-        
```
can this file can be reverted?

ashish-amarnath commented on site/docs/v1.3.2/code-standards.md:
```diff
@@ -43,7 +43,7 @@ Example:
     	corev1client "k8s.io/client-go/kubernetes/typed/core/v1"
     	corev1listers "k8s.io/client-go/listers/core/v1"
-        
+       
```
can this file can be reverted?

carlisia commented:
@ashish-amarnath I fixed your addressable suggestions, thank you! Others I added comments for consideration.

I did not fix the import grouping suggestions. I realize we talked about this way before you joined the project: we were considering cleaning up imports a low value task. But now with the linter, maybe we can do better in this department. Let's take up this convo there and figure out what we do going forward. I appreciate the attention to detail!

nrb commented:
RE: import order, let's discuss it in the linter settings on https://github.com/vmware-tanzu/velero/pull/2615. I would rather we get it set in the linter settings than debate it in each PR, or ask us to manually do the linter's work.

One thing I appreciate about Go is that we can lean on tools like `gofmt` to do the formatting for us. I think we should get the setting we want into `goimports`, then let it take care of things.

carlisia commented on pkg/apis/velero/v1/backupstoragelocation_types.go:
```diff
@@ -21,33 +21,81 @@ import (
+	AccessMode BackupStorageLocationAccessMode `json:"accessMode,omitempty"`
+}
+
+// +genclient
```
No. I've added this comment: `// TODO(2.0) After converting all resources to use the runttime-controller client,
// the genclient and k8s:deepcopy markers will no longer be needed and should be removed.`

carlisia commented on PROJECT:
```diff
@@ -0,0 +1,7 @@
+- group: velero
+  kind: BackupStorageLocation
+  version: v1
+version: "2"
```
It was defaulted to this value. I'll find out.

carlisia replied:
Answer from the `kubebuilder` k8s channel:

```
Kubebuilder's major version is 2
And tracked for any required breaking upgrade helpers in the future
```

alaypatel07 commented on pkg/controller/backup_controller.go:
```diff
@@ -57,6 +58,8 @@ import (
 	"github.com/vmware-tanzu/velero/pkg/util/logging"
 	"github.com/vmware-tanzu/velero/pkg/volume"
+
+	"sigs.k8s.io/controller-runtime/pkg/client"
```
qq: should we alias this package to kbclient in order to maintain consistency?

alaypatel07 commented on pkg/controller/backup_controller_test.go:
```diff
@@ -18,8 +18,12 @@ package controller
 	"fmt"
 	"io"
+
+	"sigs.k8s.io/controller-runtime/pkg/client"
```
please see alias comment question, thanks

alaypatel07 commented on pkg/restic/common.go:
```diff
@@ -17,11 +17,15 @@ limitations under the License.
 	"strings"
 	"time"
 
+	"sigs.k8s.io/controller-runtime/pkg/client"
```
duplicate import for the same package see line below, not sure if I am missing something but looks like a typo

ashish-amarnath commented on pkg/cmd/cli/backuplocation/get.go:
```diff
@@ -36,19 +40,24 @@ func NewGetCommand(f client.Factory, use string) *cobra.Command {
 			cmd.CheckError(err)
 
-			var locations *api.BackupStorageLocationList
+			locations := new(velerov1api.BackupStorageLocationList)
```
Here we want to create a new list object and not a slice of backupstoragelocations.
This, IMO, is correct.

carlisia commented on internal/util/managercontroller/managercontroller.go:
```diff
@@ -0,0 +1,57 @@
+	go func() {
+		select {
+		case <-parentCh:
+			cancel()
```
Not sure if you mean instead of cancel() otherwise there's no need to add a return, that's the default behavior.

carlisia replied:
I mean, it returns from the switch but we falls down to return ctx, cancel in the end. I don't think it's needed to add that here, this seems fine to me. If I misunderstood please let me know.

shomron replied:
There's no need for an explicit return - if this case is hit it will cancel the child context, exit the select statement, and then exit the goroutine.

shomron commented on internal/util/managercontroller/managercontroller.go:
```diff
@@ -0,0 +1,57 @@
+		select {
+		case <-parentCh:
+			cancel()
+		case <-ctx.Done():
```
How is this case a no-op? The goroutine's mission is to either close the child context if the parent is closed, or exit when only the child is cancelled. The child can be cancelled without the parent being closed. Without the second case - the goroutine would leak.
